### PR TITLE
net: socket: add SO_SNDTIMEO timeout option

### LINKED ISSUE #11820: sockets: Implement (POSIX-compatible) timeout support
It's not possible to implement a robust networking solution without timeouts. So, sockets should implement them, following the POSIX mandated API.

### PR DESCRIPTION
**Pull request by hubertmis**
This PR is intended to fix #11820.

Currently handling this option is implemented for sockets and partially for net context, but as pointed out in #31369, underlying net_context implementation of sending functions ignores timeout parameters. Because of that send timeout options is not functional yet.

### CODE REVIEW

**Comment by pfalcon** [comment]
@hubertmis: Sorry for not thanking you earlier for continuing the effort of https://github.com/zephyrproject-rtos/zephyr/pull/31369, and submitting this PR. As we now have pre-release freeze for the 2.5 release, people are concentrating on polishing it. We'll get to this PR after the release. Thanks!

**Comment by hubertmis** [comment]
> As we now have pre-release freeze for the 2.5 release, people are concentrating on polishing it. We'll get to this PR after the release.

No problem, I'm aware of the ongoing release. I didn't mean to take time of any of the maintainers to look at these patches now. I'm just stacking pull requests while I'm creating patches for my project. I was expecting they'll be unstacked after release, one by one.

No worries, no rush :)

This particular one requires some attention. I would not merge it in current shape. We should discuss it

**Comment by expploitt** [comment]
Hello! This PR will be available soon? Thanks!

File: subsys/net/lib/sockets/sockets.c [hunk 1]
```diff
@@ -1632,6 +1635,39 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
...
+					errno = -ret;
+					return -1;
+				}
+
+				return 0;
+			}
+
+			break;
+
+
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c:1670]: one empty line is enough here.

File: subsys/net/lib/sockets/sockets.c [hunk 2]
```diff
@@ -1632,6 +1635,39 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
...
+				} else {
+					timeout = K_USEC(tv->tv_sec * 1000000ULL
+							 + tv->tv_usec);
+				}
+
+				ret = net_context_set_option(ctx,
+							     NET_OPT_SNDTIMEO,
+							     &timeout,
+							     sizeof(timeout));
+
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c:1659]: this empty line can be removed

**Review by jukkar — COMMENTED** [review]
@hubertmis Could you fix the merge conflict and re-submit

**Review by jukkar — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**